### PR TITLE
Update v8 to include Promise Context Tagging patch

### PR DESCRIPTION
**Pull request by jasnell**
Implements a mechanism in v8 that allows tagging a Promise with a creation context, then using that to implement cross-request promise awaits.

The way the mechanism works is fairly straightforward:

1. Whenever we enter IoContext::Scope, we'll set a "promise context tag" on the Isolate, currently this is just an opaque empty object that we use only as a marker.
2. Whenever a JS Promise is created while we are in that IoContext::Scope, the Promise will be tagged with that same promise context tag object. When a JS Promise is created outside of the IoContext::Scope, its context tag is empty.
3. Whenever a JS Promise is followed (e.g. using await or by calling then, etc) the context tag on the promise is compared against the isolate's current context tag. If they are the same, or if the JS Promise's context tag is empty (meaning it was created outside of an IoContext) then nothing changes and we attach the continuation to THAT JS Promise. However, if the tags are different, that implies that we are following a JS Promise from one IoContext from a different IoContext and so our new PromiseCrossContextCallback is invoked.
4. The PromiseCrossContextCallback creates a CrossThreadWaitList and associates that with the JS Promise being followed. We attached continuations to that JS Promise that will fulfill or reject the CrossThreadWaitList.
5. A new JS Promise within the following IoContext is created and is set up to follow the CrossThreadWaitList. This new JS Promise replaces the promise that is being followed, and our then/await continuations are attached to this new Promise instead.

There are plenty of edge cases to work through to ensure that things are working but the fundamental idea appears to work with very little performance impact. I'll tag you both once I have the relevant PRs open.

The simple test case that I have been using to test that things work is simply:

```
export default {
  async fetch() {
    globalThis.promise ??= (async () => {
      console.log('fetching...');
      const resp = await fetch('https://example.com');
      return await resp.text();
    })();
    const text = await globalThis.promise;
    return new Response(`ok ${text.length}`);
  }
}
```

Then using simply curl localhost:8080 & curl localhost:8080 to queue up simultaneous requests should be enough to verify that it basically works. I was using autocannon to throw more concurrent load to test that things work with gc. If all goes well, every request should be handled successfully and the fetch should only ever happen once.

### CODE REVIEW

**Comment by jasnell** [comment]
Note: The PR still needs tests added to ensure it's working correctly. I'll likely add the tests to the internal PR since it'll be easier.

File: patches/v8/0012-Implement-Promise-Context-Tagging.patch [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+ }
+ 
++class Isolate::PromiseContextScope {
++public:
++  PromiseContextScope(Isolate* isolate, v8::Local<v8::Object> tag);
++  ~PromiseContextScope();
++  PromiseContextScope(const PromiseContextScope&) = delete;
++  PromiseContextScope(PromiseContextScope&&) = delete;
++  PromiseContextScope& operator=(const PromiseContextScope&) = delete;
++  PromiseContextScope& operator=(PromiseContextScope&&) = delete;
```
Comment (ohodson) [inline patches/v8/0012-Implement-Promise-Context-Tagging.patch:70]: V8 would typically use `DISALLOW_IMPLICIT_CONSTRUCTORS(PromiseContextScope)` for all but the move constructor here (not fanaticially used everywhere, jfyi).

Comment (ohodson) [reply patches/v8/0012-Implement-Promise-Context-Tagging.patch:70]: Overall this patch LGTM. I've run v8 tests against the new patch 0012 for release and debug builds with no issues.

File: src/workerd/io/io-context.c++ [hunk 2]
```diff
@@ -1092,7 +1092,8 @@ public:
         threadScope(context),
         workerLock(*context.worker, lockType),
         handleScope(workerLock.getIsolate()),
-        jsContextScope(workerLock.getContext()) {
+        jsContextScope(workerLock.getContext()),
+        promiseContextScope(workerLock.getIsolate(), context.getPromiseContextTag(workerLock)) {
```
Comment (ohodson) [inline src/workerd/io/io-context.c++:1096]: newline per field to match local style

**Review by ohodson — COMMENTED** [review]
LGTM, but probably best to have approval from Kenton for greater familiarity with most of the parts here.

File: src/workerd/io/worker.c++ [hunk 3]
```diff
@@ -1060,6 +1113,46 @@ Worker::Isolate::Isolate(kj::Own<ApiIsolate> apiIsolateParam,
...
+  // that scope will be tagged with the same promise context tag. When an attempt to
+  // follow a promise occurs (e.g. either using Promise.prototype.then() or await, etc)
+  // our patched v8 logic will check to see if the followed promise's tag matches the
+  // current Isolate tag. If they do not, then v8 will invoke this callback. The promise
+  // here is the promise that belongs to a different IoContext.
+  lock->v8Isolate->SetPromiseCrossContextCallback([](v8::Local<v8::Context> context,
+                                                     v8::Local<v8::Promise> promise,
+                                                     v8::Local<v8::Object> tag) ->
+                                                       v8::MaybeLocal<v8::Promise> {
+    if (!IoContext::hasCurrent()) return promise;
```
Comment (kentonv) [inline src/workerd/io/worker.c++:1129]: Maybe we should return a rejected promise saying that cross-context awaits don't work outside of a request context?

File: src/workerd/io/worker.c++ [hunk 4]
```diff
@@ -1060,6 +1113,46 @@ Worker::Isolate::Isolate(kj::Own<ApiIsolate> apiIsolateParam,
...
+
+    // Note that we capture a strong (non-traced) reference to the promise here to ensure
+    // that we can still get access to the CrossThreadPromiseWaiter so we can extract the
+    // resolved value after the waiter promise resolves. We have to do it this way because
+    // the CrossThreadWaitList used by the waiter is not currently capable of following
+    // anything but a kj::Promise<void>. We end up having to pass the value out of band.
+    // That's unfortunate but works for now.
+    // TODO(cleanup): Update CrossThreadWaitList to support passing non-void values.
+    return js.wrapSimplePromise(waiter.addWaiter(js).then(js,
+        [promise=kj::mv(promiseRef)](jsg::Lock& js) mutable -> jsg::Promise<jsg::Value> {
```
Comment (kentonv) [inline src/workerd/io/worker.c++:1146]: Can this lambda return just `jsg::Value` instead of `jsg::Promise<jsg::Value>`? It looks like all returns construct an immediately-resolved promise.

File: src/workerd/io/worker.c++ [hunk 5]
```diff
@@ -978,6 +979,58 @@ kj::Maybe<kj::String> makeCompatJson(kj::ArrayPtr<kj::StringPtr> enableFlags) {
   return kj::String(json.releaseAsArray());
 }
 
+struct CrossThreadPromiseWaiter {
+  // When a promise is created in a different IoContext, we need to use a
+  // CrossThreadWaitList in order to wait on it. The Waiter instance will
+  // be held on the Promise itself, and will be fulfilled/rejected when the
+  // promise is resolved or rejected. This will signal all of the waiters
+  // from other IoContexts.
+  CrossThreadWaitList waitList;
```
Comment (kentonv) [inline src/workerd/io/worker.c++:988]: Instead of using `CrossThreadWaitList`, it seems like the code below would be simpler if you constructed a new `kj::newPromiseAndCrossThreadFulfiller()` pair each time the promise is cross-context awaited. Then you don't have to try to coordinate with any other waiter on the same process.

IIRC `CrossThreadWaitList` is essentially doing the same thing internally (creating a new pair for each waiter) so you aren't getting any performance benefit from trying to use it here.

File: src/workerd/io/worker.c++ [hunk 6]
```diff
@@ -978,6 +979,58 @@ kj::Maybe<kj::String> makeCompatJson(kj::ArrayPtr<kj::StringPtr> enableFlags) {
...
+
+    auto waiter = kj::heap<CrossThreadPromiseWaiter>();
+    auto onSuccess = [promise=promise.addRef(js)]
+        (jsg::Lock& js, jsg::Value value) mutable {
+      getCrossThreadPromiseWaiter(js, promise).fulfill(kj::mv(value));
+    };
+
+    auto onFailure = [promise=promise.addRef(js)]
+        (jsg::Lock& js, jsg::Value exception) mutable {
+      getCrossThreadPromiseWaiter(js, promise).reject(js.exceptionToKj(kj::mv(exception)));
```
Comment (kentonv) [inline src/workerd/io/worker.c++:1023]: Converting the exception to KJ and back is unfortunate since the recipient is in the same isolate -- letting the JavaScript error object flow through would be best.

Here's an idea: Have the cross-thread promise-fulfiller be of type `<void>`. Have both the success and failure methods simply fulfill the fulfiller. On the receiving end, the `.then()` continuation should _return the original promise_. Since we know the promise is settled at this point, it's now OK to use it directly. This approach correctly propagates either a result value or an error without ever needing to store them anywhere in C++ structures.

File: src/workerd/io/worker.c++ [hunk 7]
```diff
@@ -1060,6 +1113,46 @@ Worker::Isolate::Isolate(kj::Own<ApiIsolate> apiIsolateParam,
...
+                                                     v8::Local<v8::Object> tag) ->
+                                                       v8::MaybeLocal<v8::Promise> {
+    if (!IoContext::hasCurrent()) return promise;
+    auto isolate = context->GetIsolate();
+    auto& js = jsg::Lock::from(isolate);
+    auto promiseRef = js.v8Ref(promise);
+    // The CrossThreadPromiseWaiter is a help that is attached to the promise. It is
+    // allows multiple cross-thread/cross-request waiters to be attached to the promise,
+    // all of which will be fulfilled when promise is resolved.
+    auto& waiter = getCrossThreadPromiseWaiter(js, promiseRef);
```
Comment (kentonv) [inline src/workerd/io/worker.c++:1136]: If `promise` is already resolved, then there's no need to set up a cross-thread wait. You can safely just return it. Can we check `promise->State()` and optimize here?

**Review by kentonv — COMMENTED** [review]
Is there any performance impact for workers that don't use this?

Can we write a test case? I think the easiest way to do this would be to set up a test worker with a service binding (perhaps to an alternate entrypoint of itself). The test can then initiate multiple subrequests to the service binding and on the receiving end verify that they can synchronize with each other.

File: src/workerd/io/worker.c++ (hunk 7 above)
Comment (jasnell) [reply src/workerd/io/worker.c++:1136]: If `promise` is already resolved this callback wouldn't have been called.

File: src/workerd/io/worker.c++ (hunk 3 above)
Comment (kentonv) [reply src/workerd/io/worker.c++:1129]: Although I suppose _technically_ it might be OK because if the listener isn't in a request context, then its continuation isn't going to access anything inside the request context, so it's OK to run it in the publisher's context. But it might still be confusing, and I don't know what the use case would be, so maybe better to stick with an exception for now.

File: src/workerd/io/worker.c++ (hunk 7 above)
Comment (kentonv) [reply src/workerd/io/worker.c++:1136]: Ah, great.

**Comment by jasnell** [comment]
Internal PR updated with a test.

**Review by kentonv — APPROVED** [review]
I don't understand the V8 code so will trust @ohodson's review there.
